### PR TITLE
Respect --stacktrace for additional failures logged in afterEvaluate()

### PR DESCRIPTION
**Pull request by jjohannes**
See: #874

As prerequisite, this PR contains:
- Fix for failure handling in BuildOperationExecutor
- Adding `withStacktraceDisabled()` to test executors

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java [hunk 1]
```diff
@@ -76,7 +77,13 @@ private void notifyAfterEvaluate(ProjectEvaluationListener listener, ProjectInte
         } catch (Exception e) {
             if (state.hasFailure()) {
                 // Just log this failure, and pass the existing failure out in the project state
-                LOGGER.error("Failed to notify ProjectEvaluationListener.afterEvaluate(), but primary configuration failure takes precedence.", e);
+                boolean logStackTraces = project.getGradle().getStartParameter().getShowStacktrace() != ShowStacktrace.INTERNAL_EXCEPTIONS;
+                String infoMessage = "Failed to notify ProjectEvaluationListener.afterEvaluate(), but primary configuration failure takes precedence.";
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: ❌ We should reword this too, because users mistake this for the real error. How about:

`Project evaluation failed, leading to a follow-up error in Project.afterEvaluate {}. Run with --stacktrace to see the follow-up error`.

File: subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java [hunk 2]
```diff
@@ -76,7 +77,13 @@ private void notifyAfterEvaluate(ProjectEvaluationListener listener, ProjectInte
...
+                if (logStackTraces) {
+                    LOGGER.error(infoMessage, e);
+                } else {
+                    LOGGER.error(infoMessage + "\n> {}", e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName());
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: Maybe we don't even print the message, but only the info above.

File: subprojects/core/src/test/groovy/org/gradle/internal/progress/DefaultBuildOperationExecutorTest.groovy [hunk 3]
```diff
@@ -651,6 +651,10 @@ class DefaultBuildOperationExecutorTest extends ConcurrentSpec {
         def e = thrown(ResourceDeadlockException)
     }
 
+    def "Writes exception to "() {
```
Comment (oehme) [inline subprojects/core/src/test/groovy/org/gradle/internal/progress/DefaultBuildOperationExecutorTest.groovy]: ? :)

File: subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java (hunk 1 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: Good suggestion.

File: subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java (hunk 2 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: Yes... and you would see the message as primary failure after you fixed the other one.

File: subprojects/core/src/test/groovy/org/gradle/internal/progress/DefaultBuildOperationExecutorTest.groovy (hunk 3 above)
Comment (jjohannes) [inline subprojects/core/src/test/groovy/org/gradle/internal/progress/DefaultBuildOperationExecutorTest.groovy]: Friday... 😐

File: subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java (hunk 2 above)
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: In most cases the secondary failure is *because of* the primary failure. So once you fix the primary one, the other will disappear as well. That's why I think it's good to hide it.

File: subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java (hunk 1 above)
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: I was about to add a comment about the error message. I am bit unsure what we want to express here with "leading to a follow-up error". 

- There are simply two different errors during project configuration: one outside of `afterEvaluate` and one inside.
- Because project evaluation failed the exception in `afterEvaluate` is triggered as well.

As a user I don't think I need to know these implementation details. What I want to know is that there are two issues.

Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: Yes that's right. It does not have to be a follow-up error in all cases.

How about something like:

`Project evaluation failed including an error in afterEvaluate {}. Run with --stacktrace for details of the afterEvaluate {} error.`

Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: I think it's a clearer representation of what's going on for users. I'd vote to go with that.

Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/configuration/project/LifecycleProjectEvaluator.java]: I changed it.
